staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go

```diff
@@ -618,6 +541,92 @@ func (td *typeDiscoverer) getValidationFunctionName(t *types.Type) (types.Name,
 	return types.Name{Package: pkg, Name: "Validate_" + t.Name.Name}, true
 }
 
+// emitRegisterFunction emits the type-registration logic for validation
+// functions.
+func (g *genValidations) emitRegisterFunction(c *generator.Context, sw *generator.SnippetWriter) {
+	scheme := c.Universe.Type(schemeType)
+	schemePtr := &types.Type{
+		Kind: types.Pointer,
+		Elem: scheme,
+	}
+
+	sw.Do("func init() { localSchemeBuilder.Register(RegisterValidations)}\n\n", nil)
+
+	sw.Do("// RegisterValidations adds validation functions to the given scheme.\n", nil)
+	sw.Do("// Public to allow building arbitrary schemes.\n", nil)
+	sw.Do("func RegisterValidations(scheme $.|raw$) error {\n", schemePtr)
+	for _, t := range g.rootTypes {
+		node := g.discovered.typeNodes[t]
+		if node == nil {
+			// Should never happen.
+			klog.Fatalf("found nil node for root-type %v", t)
+		}
+
+		// TODO: It would be nice if these were not hard-coded.
```

Comment:
Having a less huristical way to identifying if a resource is "flat" or has status resource would be really nice.  As a follow up, maybe we put the huristic somewhere in our discovery code along with some comments?

Reply:
Either heuristic or explicit -- there could be a tag like `//+ subresource=status` on the status field.